### PR TITLE
CI: Initial additions for fp32 and windows GPU test support

### PR DESCRIPTION
**Pull request by ethanglaser**
# Description
Changes for green CI on GPU validation on devices without fp64 support (includes windows GPU sklearnex validation)

To be merged with infra PR 713

Changes include:

- Correct sequencing of _convert_to_supported() function for proper dtype
- Use of result.dtype instead of pytest parameter dtype so that in non fp32-supported systems where fp64 is passed into the estimator, the necessary tolerance is selected
- Minor general threshold realignments based on initial test results
- Disabling of forest GPU tests as results are inconsistent, task create to look further into the issue
- Add required platform for spmd algo examples (spmd backend not created for windows)

### CODE REVIEW

**Comment by icfaust** [comment]
@ethanglaser PRs #1674 should solve some of the outstanding issues with IncrementalEmpiricalCovariance for fp32 sklearnex testing, but it needs #1795 to be pulled in first.  I will make a meeting for #1674, but could you review #1795? Its a relatively small change, already passes CI

**Comment by ethanglaser** [comment]
Latest CI with just GPU tests: http://intel-ci.intel.com/ef0c1101-c995-f106-9784-a4bf010d0e2e

File: onedal/cluster/tests/test_kmeans_init.py [hunk 1]
```diff
@@ -84,7 +84,11 @@ def test_generated_dataset(queue, dtype, n_dim, n_cluster):
         nn = NearestNeighbors(n_neighbors=1)
         d, i = nn.fit(rs_centroids).kneighbors(cs)
         # We have applied 2 sigma rule once
-        desired_accuracy = int(0.9973 * n_cluster)
+        desired_accuracy = (
+            int(0.9973 * n_cluster) - 1
+            if d.dtype == np.float64
+            else int(0.9973 * n_cluster)
+        )
```
Comment (samir-nasibli) [inline onedal/cluster/tests/test_kmeans_init.py:91]:
```suggestion
        desired_accuracy = int(0.9973 * n_cluster)
        if d.dtype == np.float64:
            desired_accuracy = desired_accuracy - 1
```

File: onedal/linear_model/incremental_linear_model.py [hunk 2]
```diff
@@ -77,22 +77,18 @@ def partial_fit(self, X, y, queue=None):
...
 
-        if self._dtype not in [np.float32, np.float64]:
-            self._dtype = np.float64
-
-        X = X.astype(self._dtype, copy=self.copy_X)
-        y = y.astype(dtype=self._dtype)
+        y = np.asarray(y).astype(dtype=self._dtype)
         self._y_ndim_1 = y.ndim == 1
 
-        X, y = _check_X_y(X, y, force_all_finite=False, accept_2d_y=True)
```
Comment (icfaust) [inline onedal/linear_model/incremental_linear_model.py:91]: force_all_finite occurs here for partial fit (https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/linear_model/incremental_linear.py#L157) Ideally we would disable the finite check there to allow for a single finite check.

File: onedal/linear_model/tests/test_logistic_regression.py [hunk 3]
```diff
@@ -38,6 +38,13 @@ def test_breast_cancer(queue, dtype):
         model = LogisticRegression(fit_intercept=True, solver="newton-cg")
         model.fit(X_train, y_train, queue=queue)
         y_pred = model.predict(X_test, queue=queue)
+
+        # TODO: check why predictions all the same on windows GPU
```
Comment (icfaust) [inline onedal/linear_model/tests/test_logistic_regression.py:42]: Send me a link to the ticket associated with the TODO, and remove the TODO.

File: onedal/primitives/tests/test_kernel_functions.py [hunk 4]
```diff
@@ -66,7 +66,8 @@ def test_dense_self_rbf_kernel(queue):
     result = rbf_kernel(X, queue=queue)
     expected = sklearn_rbf_kernel(X)
 
-    assert_allclose(result, expected, rtol=1e-14)
+    tol = 1e-5 if result.dtype == np.float32 else 1e-14
```
Comment (icfaust) [inline onedal/primitives/tests/test_kernel_functions.py:69]: Yikes, a 10^9 change in performance. Would this also warrant an investigation?

File: sklearnex/ensemble/tests/test_forest.py [hunk 5]
```diff
@@ -45,7 +46,10 @@ def test_sklearnex_import_rf_classifier(dataframe, queue):
     assert_allclose([1], _as_numpy(rf.predict([[0, 0, 0, 0]])))
 
 
-@pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+# TODO: fix RF regressor predict for the GPU sycl_queue.
```
Comment (icfaust) [inline sklearnex/ensemble/tests/test_forest.py:49]: Similar comment about ticket instead of todo

File: sklearnex/ensemble/tests/test_forest.py [hunk 6]
```diff
@@ -65,7 +69,10 @@ def test_sklearnex_import_rf_regression(dataframe, queue):
             assert_allclose([-6.839], pred, atol=1e-2)
 
 
-@pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+# TODO: fix ET classifier predict for the GPU sycl_queue.
```
Comment (icfaust) [inline sklearnex/ensemble/tests/test_forest.py:72]: Ticket here as well.

**Review by icfaust — COMMENTED** [review]
Just some small things, otherwise looks fine.

File: onedal/primitives/tests/test_kernel_functions.py (hunk 4 above)
Comment (ethanglaser) [reply onedal/primitives/tests/test_kernel_functions.py:69]: I am not sure why 1e-14 was used here specifically - generally we don't go more specific than 1e-7. For rtol, 1e-5 should be acceptable, as this is a typical threshold used in other fp32 testing

File: onedal/linear_model/incremental_linear_model.py (hunk 2 above)
Comment (icfaust) [reply onedal/linear_model/incremental_linear_model.py:91]: Recommended change, add force_all_finite=False to https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/linear_model/incremental_linear.py#L157, https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/linear_model/incremental_linear.py#L166 and https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/linear_model/incremental_linear.py#L171

**Comment by ethanglaser** [comment]
Should be ready for final review

File: onedal/linear_model/tests/test_logistic_regression.py (hunk 3 above)
Comment (ethanglaser) [reply onedal/linear_model/tests/test_logistic_regression.py:42]: hopefully resolved by https://github.com/intel/scikit-learn-intelex/pull/1851, removing skip to check

**Comment by ethanglaser** [comment]
CI with infra branch (includes fp32 + windows GPU validation): http://intel-ci.intel.com/ef2436e8-ef5e-f182-b1e2-a4bf010d0e2e

File: onedal/cluster/tests/test_kmeans_init.py [hunk 7]
```diff
@@ -85,6 +85,8 @@ def test_generated_dataset(queue, dtype, n_dim, n_cluster):
         d, i = nn.fit(rs_centroids).kneighbors(cs)
         # We have applied 2 sigma rule once
         desired_accuracy = int(0.9973 * n_cluster)
+        if d.dtype == np.float64:
+            desired_accuracy = desired_accuracy - 1
```
Comment (md-shafiul-alam) [inline onedal/cluster/tests/test_kmeans_init.py:89]: What was the logic behind the desired accuracy -1? Will it work for all values of n_cluster?

Comment (ethanglaser) [reply onedal/cluster/tests/test_kmeans_init.py:89]: it matches a minor threshold change added to test_kmeans (https://github.com/intel/scikit-learn-intelex/blob/main/onedal/cluster/tests/test_kmeans.py#L87) not sure exactly how the desired_accuracy was set but seems the threshold doesn't necessarily matchup with actual performance

**Review by md-shafiul-alam — APPROVED** [review]
looks good to me, but worth waiting for @icfaust about all the force_all_finite business

File: onedal/cluster/tests/test_kmeans_init.py (hunk 7 above)
Comment (md-shafiul-alam) [reply onedal/cluster/tests/test_kmeans_init.py:89]: I see, the kmeans++ init issue may be get fixed with the changes proposed here https://github.com/oneapi-src/oneDAL/pull/2796

**Review by icfaust — APPROVED** [review]

**Comment by ethanglaser** [comment]
Thanks for reviews :)
